barter-data/src/exchange/coinbase/book/l1.rs

```diff
@@ -0,0 +1,177 @@
+use crate::{
+    event::{MarketEvent, MarketIter},
+    exchange::{coinbase::channel::CoinbaseChannel, subscription::ExchangeSub, ExchangeId},
+    books::Level,
+    subscription::book::OrderBookL1,
+    Identifier,
+};
+use barter_integration::subscription::SubscriptionId;
+use chrono::{DateTime, Utc};
+use rust_decimal::Decimal;
+use serde::{Deserialize, Serialize};
+
+/// [`Coinbase`](super::super::Coinbase) real-time OrderBook Level1 (top of book) message.
+///
+/// ### Raw Payload Examples
+/// #### Coinbase OrderBookL1
+/// See docs: <https://docs.cdp.coinbase.com/exchange/docs/websocket-channels/#ticker-channel>
+/// ```json
+///{
+///  "type": "ticker",
+///  "product_id": "ETH-USD",
+///  "price": "1285.22",
+///  "open_24h": "1310.79",
+///  "volume_24h": "245532.79269678",
+///  "low_24h": "1280.52",
+///  "high_24h": "1313.8",
+///  "volume_30d": "9788783.60117027",
+///  "best_bid": "1285.04",
+///  "best_bid_size": "0.46688654",
+///  "best_ask": "1285.27",
+///  "best_ask_size": "1.56637040",
+///  "side": "buy",
+///  "time": "2022-10-19T23:28:22.061769Z",
+///  "trade_id": 370843401,
+///  "last_size": "11.4396987"
+///}
+/// ```
+#[derive(Clone, PartialEq, PartialOrd, Debug, Deserialize, Serialize)]
+pub struct CoinbaseOrderBookL1 {
+    #[serde(rename(deserialize = "time"))]
+    pub time: DateTime<Utc>,
+    #[serde(rename(deserialize = "type"))]
+    pub kind: String,
+    #[serde(
+        rename(deserialize = "product_id"),
+        deserialize_with = "de_ob_l1_subscription_id"
+    )]
+    pub subscription_id: SubscriptionId,
+    #[serde(
+        rename(deserialize = "best_bid"),
+        deserialize_with = "barter_integration::de::de_str"
+    )]
+    pub best_bid_price: Decimal,
+    #[serde(
+        rename(deserialize = "best_bid_size"),
+        with = "rust_decimal::serde::str"
+    )]
+    pub best_bid_amount: Decimal,
+    #[serde(
+        rename(deserialize = "best_ask"),
+        with = "rust_decimal::serde::str"
+    )]
+    pub best_ask_price: Decimal,
+    #[serde(
+        rename(deserialize = "best_ask_size"),
+        with = "rust_decimal::serde::str"
+    )]
+    pub best_ask_amount: Decimal,
+}
+
+impl Identifier<Option<SubscriptionId>> for CoinbaseOrderBookL1 {
+    fn id(&self) -> Option<SubscriptionId> {
+        Some(self.subscription_id.clone())
+    }
+}
+
+impl<InstrumentId> From<(ExchangeId, InstrumentId, CoinbaseOrderBookL1)>
```

Comment:
You can drop the ExchangeId from the tuple since we already know it's `ExchangeId::Coinbase`. I know this isn't the case in other places, but let's not add more occurances :)

Reply:
Ah, I see why now... the StatelessTransformer bound of:

`MarketIter<InstrumentKey, Kind::Event>: From<(ExchangeId, InstrumentKey, Input)>`

Okay, let's keep that for now!